src/libstd/io/mod.rs

```diff
@@ -967,18 +967,23 @@ impl Initializer {
 ///
 /// Implementors of the `Write` trait are sometimes called 'writers'.
 ///
-/// Writers are defined by two required methods, [`write`] and [`flush`]:
+/// Writers are intended to be composable with one another. Many implementors
+/// throughout [`std::io`] take and provide types which implement the `Write`
+/// trait. For example, `BufWriter` may be composed with `TcpStream`,
+/// buffering the output passed to its `write` method, until its internal
+/// buffer is full, or the `flush` method is invoked.
+///
+/// The 'true sink' is the final writer in a chain of writers, for example,
+/// `TcpStream`, `File`, or `Cursor`. For a true sink the ['flush']
+/// operation is a no-op which always returns `Ok()`.
```

Comment:
I don't think this is accurate: For example, AFAIK, writing to a file may succeed, however data is only actually written to a kernerl-internal buffer. Calling flush ensures that this buffer is actually written to disk, which may still produce an error.

Stating that flush on a "true sink" will never fail seems like a pretty strong guarantee, which I don't think we should make.

Reply:
I spend a fair amount of time examining the source code of the standard library in the current rust repo. What you are maintaining occurs is exactly what I imagined occurred, until inspection and observation seemed to indicate differently. I think this is an _important_ point, which is why I filed the PR.

Here's some code taken from the OS specific implementations of various things. You will note that in almost all of them, ```flush()``` just returns ```Ok()```. In the only one that's different, the ```flush()``` is just forwarded to a ```Stderr::flush()``` or ```Stdout::flush()``` which do just return ```Ok()```.

```
[mulhern@localhost rust]$ rg 'fn flush' --context 5 src/libstd/sys
src/libstd/sys/unix/stdio.rs
35-        let ret = fd.write(data);
36-        fd.into_raw();
37-        ret
38-    }
39-
40:    pub fn flush(&self) -> io::Result<()> {
41-        Ok(())
42-    }
43-}
44-
45-impl Stderr {
--
50-        let ret = fd.write(data);
51-        fd.into_raw();
52-        ret
53-    }
54-
55:    pub fn flush(&self) -> io::Result<()> {
56-        Ok(())
57-    }
58-}
59-
60-// FIXME: right now this raw stderr handle is used in a few places because
--
63-impl io::Write for Stderr {
64-    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
65-        Stderr::write(self, data)
66-    }
67-
68:    fn flush(&mut self) -> io::Result<()> {
69-        Stderr::flush(self)
70-    }
71-}
72-
73-pub fn is_ebadf(err: &io::Error) -> bool {

src/libstd/sys/unix/fs.rs
506-
507-    pub fn write_at(&self, buf: &[u8], offset: u64) -> io::Result<usize> {
508-        self.0.write_at(buf, offset)
509-    }
510-
511:    pub fn flush(&self) -> io::Result<()> { Ok(()) }
512-
513-    pub fn seek(&self, pos: SeekFrom) -> io::Result<u64> {
514-        let (whence, pos) = match pos {
515-            // Casting to `i64` is fine, too large values will end up as
516-            // negative which will cause an error in `lseek64`.

src/libstd/sys/cloudabi/stdio.rs
35-            io::ErrorKind::BrokenPipe,
36-            "Stdout is not connected to any output in this environment",
37-        ))
38-    }
39-
40:    pub fn flush(&self) -> io::Result<()> {
41-        Ok(())
42-    }
43-}
44-
45-impl Stderr {
--
52-            io::ErrorKind::BrokenPipe,
53-            "Stderr is not connected to any output in this environment",
54-        ))
55-    }
56-
57:    pub fn flush(&self) -> io::Result<()> {
58-        Ok(())
59-    }
60-}
61-
62-// FIXME: right now this raw stderr handle is used in a few places because
--
65-impl io::Write for Stderr {
66-    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
67-        Stderr::write(self, data)
68-    }
69-
70:    fn flush(&mut self) -> io::Result<()> {
71-        Stderr::flush(self)
72-    }
73-}
74-
75-pub fn is_ebadf(err: &io::Error) -> bool {

src/libstd/sys/windows/fs.rs
331-
332-    pub fn write_at(&self, buf: &[u8], offset: u64) -> io::Result<usize> {
333-        self.handle.write_at(buf, offset)
334-    }
335-
336:    pub fn flush(&self) -> io::Result<()> { Ok(()) }
337-
338-    pub fn seek(&self, pos: SeekFrom) -> io::Result<u64> {
339-        let (whence, pos) = match pos {
340-            // Casting to `i64` is fine, `SetFilePointerEx` reinterprets this
341-            // integer as `u64`.

src/libstd/sys/redox/stdio.rs
35-        let ret = fd.write(data);
36-        fd.into_raw();
37-        ret
38-    }
39-
40:    pub fn flush(&self) -> io::Result<()> {
41-        cvt(syscall::fsync(1)).and(Ok(()))
42-    }
43-}
44-
45-impl Stderr {
--
50-        let ret = fd.write(data);
51-        fd.into_raw();
52-        ret
53-    }
54-
55:    pub fn flush(&self) -> io::Result<()> {
56-        cvt(syscall::fsync(2)).and(Ok(()))
57-    }
58-}
59-
60-// FIXME: right now this raw stderr handle is used in a few places because
--
63-impl io::Write for Stderr {
64-    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
65-        Stderr::write(self, data)
66-    }
67-
68:    fn flush(&mut self) -> io::Result<()> {
69-        Stderr::flush(self)
70-    }
71-}
72-
73-pub fn is_ebadf(err: &io::Error) -> bool {

src/libstd/sys/windows/stdio.rs
159-
160-    pub fn write(&self, data: &[u8]) -> io::Result<usize> {
161-        write(c::STD_OUTPUT_HANDLE, data)
162-    }
163-
164:    pub fn flush(&self) -> io::Result<()> {
165-        Ok(())
166-    }
167-}
168-
169-impl Stderr {
--
173-
174-    pub fn write(&self, data: &[u8]) -> io::Result<usize> {
175-        write(c::STD_ERROR_HANDLE, data)
176-    }
177-
178:    pub fn flush(&self) -> io::Result<()> {
179-        Ok(())
180-    }
181-}
182-
183-// FIXME: right now this raw stderr handle is used in a few places because
--
186-impl io::Write for Stderr {
187-    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
188-        Stderr::write(self, data)
189-    }
190-
191:    fn flush(&mut self) -> io::Result<()> {
192-        Stderr::flush(self)
193-    }
194-}
195-
196-impl Output {

src/libstd/sys/wasm/stdio.rs
33-    pub fn write(&self, data: &[u8]) -> io::Result<usize> {
34-        WriteSysCall::perform(1, data);
35-        Ok(data.len())
36-    }
37-
38:    pub fn flush(&self) -> io::Result<()> {
39-        Ok(())
40-    }
41-}
42-
43-impl Stderr {
--
48-    pub fn write(&self, data: &[u8]) -> io::Result<usize> {
49-        WriteSysCall::perform(2, data);
50-        Ok(data.len())
51-    }
52-
53:    pub fn flush(&self) -> io::Result<()> {
54-        Ok(())
55-    }
56-}
57-
58-impl io::Write for Stderr {
59-    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
60-        (&*self).write(data)
61-    }
62:    fn flush(&mut self) -> io::Result<()> {
63-        (&*self).flush()
64-    }
65-}
66-
67-pub const STDIN_BUF_SIZE: usize = 0;

src/libstd/sys/unix/ext/net.rs
580-impl io::Write for UnixStream {
581-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
582-        io::Write::write(&mut &*self, buf)
583-    }
584-
585:    fn flush(&mut self) -> io::Result<()> {
586-        io::Write::flush(&mut &*self)
587-    }
588-}
589-
590-#[stable(feature = "unix_socket", since = "1.10.0")]
591-impl<'a> io::Write for &'a UnixStream {
592-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
593-        self.0.write(buf)
594-    }
595-
596:    fn flush(&mut self) -> io::Result<()> {
597-        Ok(())
598-    }
599-}
600-
601-#[stable(feature = "unix_socket", since = "1.10.0")]

src/libstd/sys/wasm/fs.rs
212-
213-    pub fn write(&self, _buf: &[u8]) -> io::Result<usize> {
214-        match self.0 {}
215-    }
216-
217:    pub fn flush(&self) -> io::Result<()> {
218-        match self.0 {}
219-    }
220-
221-    pub fn seek(&self, _pos: SeekFrom) -> io::Result<u64> {
222-        match self.0 {}

src/libstd/sys/redox/fs.rs
287-
288-    pub fn write(&self, buf: &[u8]) -> io::Result<usize> {
289-        self.0.write(buf)
290-    }
291-
292:    pub fn flush(&self) -> io::Result<()> { Ok(()) }
293-
294-    pub fn seek(&self, pos: SeekFrom) -> io::Result<u64> {
295-        let (whence, pos) = match pos {
296-            // Casting to `i64` is fine, too large values will end up as
297-            // negative which will cause an error in `lseek64`.

src/libstd/sys/cloudabi/shims/fs.rs
210-
211-    pub fn write(&self, _buf: &[u8]) -> io::Result<usize> {
212-        match self.0 {}
213-    }
214-
215:    pub fn flush(&self) -> io::Result<()> {
216-        match self.0 {}
217-    }
218-
219-    pub fn seek(&self, _pos: SeekFrom) -> io::Result<u64> {
220-        match self.0 {}
```

If you inspect the implementation of the abstract, non OS-specific ```File``` in ```src/libstd/fs.rs``` you can see that in all implementations of ```Write``` the ```flush``` call is just forwarded to the underlying implementation, i.e., all those no-ops in the OS-specific implementations.

Here's a look at the visible implementations of ```Write``` for ```TcpStream```.

```
#[stable(feature = "rust1", since = "1.0.0")]
impl Write for TcpStream {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> { self.0.write(buf) }
    fn flush(&mut self) -> io::Result<()> { Ok(()) }
}
```

and

```
#[stable(feature = "rust1", since = "1.0.0")]
impl<'a> Write for &'a TcpStream {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> { self.0.write(buf) }
    fn flush(&mut self) -> io::Result<()> { Ok(()) }
}
```

again, that it is a no-op seems pretty irrefutable.

On the other hand, ```BufWriter``` is a non-"true sink", that is, it always wraps something else that implements ```Write```. Its implementation of ```flush``` first calls ```flush_buf```, which is expected to flush all the data from its internal buffer into its contained writer, then it calls flush on that writer. And so forth.

According to my reading of the code, my explanation is the correct one, and the previous explanation was very misleading. This distinction, is of course, an important one, because, for example, developers of applications that write to files, often wish to ensure that their data is written, and it is helpful for them to known that ```flush()``` will have no effect on whether their data is actually flushed to the disk.

 ```File``` does provide to methods  ```sync_all()``` and ```sync_data()``` which have an effect on whether data is actually written to a device.

Reply:
Thank you for the investigation. It looks like, except for Redox, `flush()` is indeed a no-op for a "true sink". However I don't think we should guarantee that `flush()` on true sinks will always be a no-op. (There may be library crates out there that provide "true sinks" but can return errors from `flush()`.

cc @rust-lang/libs @rust-lang/docs: What do you think about the proposed updated `flush()` docs?

Reply:
Yes I don't think we'll want to make any hard guarantees about things like `TcpStream::flush` *always* succeeding. I think we basically just need to say that `flush` is intended to flush a sink, recursively so if it's built from smaller instances of `Write`

Reply:
What is the purpose of "redox" module?

Reply:
@mukilan support for https://github.com/redox-os

Reply:
> My primary concern was that the fact that `File::flush()` does nothing was a bit of an unpleasant surprise to me, and not a fact that the existing documentation communicates clearly.
>
> The current text describing what `File::flush` does in docs [...] is definitely misleading.

To me this rather looks like something that needs to be fixed in the `File` documentation. Especially since, while the statements that `flush()` for "true sinks" is generally a no-op may be true for `std`, it isn't necessarily true for third-party crates.

Reply:
> if it's possible to override trait docs in the impl of the trait

Yes you can. If you document an item in the impl, it will "shadow" the summary docs that would otherwise appear there.

Reply:
It's ok to decide that you don't want to guarantee behaviour of the "true sink", so long as you, by "you" I mean Rust standard library implementors, understand the consequences of that. But the entire standard library has an invariant, currently, which my doc changes merely called out (and yes, I missed the redox implementation, which broke that invariant). Right now, the docs are quite misleading, which I feel is a problem because I was mislead by them, rather significantly.

I'll take another stab at a doc fix, in a bit.

But I would like to know why the standard library implementors in this area almost all decided to make the "true sink" do nothing.